internal/messagequeue/messagequeue.go

```diff
@@ -663,7 +784,28 @@ FINISH:
 	}
 	mq.wllock.Unlock()
 
-	return mq.msg
+	// When the message has been sent, record the time at which each want was
+	// sent so we can calculate message latency
+	onSent := func() {
+		now := time.Now()
+
+		mq.wllock.Lock()
+		defer mq.wllock.Unlock()
+
+		for _, e := range peerEntries[:sentPeerEntries] {
+			if e.Cid.Defined() { // Check if want was cancelled in the interim
+				mq.peerWants.SentAt(e.Cid, now)
```

Comment:
What if this happens after we receive the block? Is that an issue?

Reply:
It should be ok - when the Session receives a block it sends cancel, and cancel removes the want from the sent wantlist. `SentAt()` checks that the sent wantlist still contains the want before recording the time a response was received:
```
func (r *recallWantlist) SentAt(c cid.Cid, at time.Time) {
	// The want may have been cancelled in the interim
	if _, ok := r.sent.Contains(c); ok {
		if _, ok := r.sentAt[c]; !ok {
			r.sentAt[c] = at
		}
	}
}
```

Reply:
Sounds right.